### PR TITLE
Faster accumulators

### PR DESCRIPTION
**Pull request by MichaReiser**
This PR implements faster accumulators. 

Accumulators are relatively costly today because collecting the accumulated values requires iterating over the queries' entire dependency tree. 
This cost is negligible with a cold DB but adds up for very deep trees. 

For example, adding accumulators to Ruff resulted in a 30% perf regression https://github.com/astral-sh/ruff/pull/14116

This PR adds a new field to `AccumulatedMap` that tracks whether any dependency of the current query has any accumulated values. It checks that field when collecting the accumulated values, allowing it to skip entire subtrees. 

This approach should make accumulated values nearly "free" in the common cases where there are no or only very few diagnostics. 

## Test Plan

I added a benchmark and I verified that the perf regression in the linked Ruff PR is mostly gone after using this salsa version.

Running the benchmark locally gives me

```
     Running benches\accumulator.rs (target\release\deps\accumulator-51d27e7934278de6.exe)
accumulator             time:   [666.27 µs 676.29 µs 688.11 µs]
                        change: [-32.104% -30.905% -29.682%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 12 outliers among 100 measurements (12.00%)
  1 (1.00%) high mild
  11 (11.00%) high severe
```

where the baseline is main

### CODE REVIEW

File: src/accumulator/accumulated_map.rs
Comment (MichaReiser) [inline src/accumulator/accumulated_map.rs:26]: I don't like the naming at all but struggled to come up with something better.

File: src/accumulator/accumulated_map.rs
Comment (davidbarsky) [inline src/accumulator/accumulated_map.rs:14]: Small nit: consider making this an enum instead? It'd make the callsites a bit more legible.

**Review by davidbarsky — APPROVED** [review]
approving modulo one nit. I'll let you decide whether to merge or not.

File: src/accumulator/accumulated_map.rs
Comment (MichaReiser) [reply src/accumulator/accumulated_map.rs]: This is a great suggestion. Thanks

**Pull request merged**
